### PR TITLE
enhance utility of extracting subgraph

### PR DESCRIPTION
**Pull request by lucienwang1009**

### CODE REVIEW

File: tools/tf_graph_tool.py [hunk 1]
```diff
@@ -167,23 +169,121 @@ def print_graph_stat(input_path):
             logging.info("\t%s = %s", op, count)
 
     @staticmethod
-    def extract_sub_graph(input_path, output_path=None, dest_nodes=None):
-        if not output_path:
-            output_path = append_file_name_suffix(input_path, "sub")
-
+    def extract_sub_graph(input_path, dest_nodes=None, output_path=None,
+                          src_nodes=None, name_prefix=None):
```
Comment (zhijxu-MS) [inline tools/tf_graph_tool.py]: add comments to explain the pamam, this function look quite long?

File: tools/tf_graph_tool.py [hunk 2]
```diff
@@ -167,23 +169,121 @@ def print_graph_stat(input_path):
...
+                    dest_nodes.append(node.name)
+        if not src_nodes:
+            src_nodes = []
+
+        if not isinstance(dest_nodes, list):
+            raise TypeError("dest_nodes must be a list.")
+        if not isinstance(src_nodes, list):
+            raise TypeError("src_nodes must be a list.")
+
+        def node_name(n):
```
Comment (zhijxu-MS) [inline tools/tf_graph_tool.py]: same with get_node_name?

File: tools/tf_graph_tool.py [hunk 3]
```diff
@@ -167,23 +169,121 @@ def print_graph_stat(input_path):
...
+            for node in graph_def.node:
+                n = node_name(node.name)
+                name_to_node[n] = node
+                name_to_input_name[n] = [node_name(x) for x in node.input]
+                name_to_seq_num[n] = seq
+                seq += 1
+            return name_to_input_name, name_to_node, name_to_seq_num
+
+
+        def assert_nodes_are_present(name_to_node, nodes):
```
Comment (zhijxu-MS) [inline tools/tf_graph_tool.py]: optional: using set.issubset?

File: tools/tf_graph_tool.py (hunk 1 above)
Comment (zhijxu-MS) [reply tools/tf_graph_tool.py]: also please explain the main logic to extract subgraph.

File: tools/tf_graph_tool.py [hunk 4]
```diff
@@ -167,23 +169,121 @@ def print_graph_stat(input_path):
...
+                if n in nodes_to_keep:
+                    # Already visited this node.
+                    continue
+                if not checker or checker(n):
+                    nodes_to_keep.add(n)
+                    next_to_visit += name_to_input_name[n]
+            return nodes_to_keep
 
-        graph_def = graph_util.extract_sub_graph(graph_def, dest_nodes)
+        if not output_path:
```
Comment (zhijxu-MS) [inline tools/tf_graph_tool.py]: put these two statements to line 284?

File: tools/tf_graph_tool.py [hunk 5]
```diff
@@ -167,23 +169,121 @@ def print_graph_stat(input_path):
...
         logging.info("load from %s", input_path)
         graph_def = load_graph_def_from_pb(input_path)
         logging.info("\ttotal node = %s", len(graph_def.node))
 
-        if dest_nodes:
-            dest_nodes = dest_nodes.split(',')
+        if (dest_nodes or src_nodes) and name_prefix:
+            raise RuntimeError("dest_nodes/src_nodes and name_prefix are incompatible.")
+        if not name_prefix:
+            name_prefix = ""
```
Comment (zhijxu-MS) [inline tools/tf_graph_tool.py]: why not make default name_prefix value to be "" instead of None?

File: tools/tf_graph_tool.py [hunk 6]
```diff
@@ -167,23 +169,121 @@ def print_graph_stat(input_path):
...
         else:
-            _, dest_nodes = get_graph_def_io_nodes(graph_def)
+            dest_nodes = []
+            for node in graph_def.node:
+                if node.name.startswith(name_prefix):
+                    dest_nodes.append(node.name)
+        if not src_nodes:
+            src_nodes = []
+
+        if not isinstance(dest_nodes, list):
```
Comment (zhijxu-MS) [inline tools/tf_graph_tool.py]: utils.make_sure?

Comment (lucienwang1009) [reply tools/tf_graph_tool.py]: don't want to use converter's utilities.

**Pull request merged**